libvast/vast/transform_steps/select.hpp

```diff
@@ -0,0 +1,29 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2021 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include "vast/expression.hpp"
+#include "vast/transform.hpp"
+
+namespace vast {
+
+// Selects mathcing rows from the input
+class select_step : public generic_transform_step {
+public:
+  select_step(std::string expr);
+
+  caf::expected<table_slice> operator()(table_slice&& slice) const override;
+
+private:
+  std::string unparsed_;
+  mutable expression expression_;
+  mutable bool parsed_;
```

Comment:
`mutable` usually implies thread-safe access. I'm not sure this is a solution you can feasibly take here. While it works _now_ because the `select_step` itself is only called from a single thread, it's really a bad API because it makes the wrong promises.

Can't we store a `caf::expected<expression>` instead and parse/normalize/tailor in the constructor, and then when applying the check we only need to check whether the expected holds a value (or return its error)?